lib/AnalysisStructured/PtrAnalysis.cpp

```diff
@@ -0,0 +1,1128 @@
+    map.map(op.getResult(), maketptrOp.getResult());
+  } else {
+    map.map(op.getResult(), op.getResult());
+  }
+  return success();
+}
+
+LogicalResult PtrAnalysis::rewriteForOp(scf::ForOp op) {
+  SmallVector<Value> newInitArgs;
+
+  SmallVector<std::pair<int, PtrSState>, 5> initArgIndexState;
+  SmallVector<std::pair<int, PtrSState>, 5> knownPtrsTmp;
+
+  llvm::SmallDenseMap<int, PtrSState> initArgIndexMap;
+
+  OpBuilder builder(op);
+
+  // Create a new list of init args
+  for (auto [i, arg] : llvm::enumerate(op.getInitArgs())) {
+    auto mappedV = map.lookupOrNull(arg);
+    PtrSState state;
+
+    if (mappedV) {
+      if (auto makeTPtrOp = mappedV.getDefiningOp<tts::MakeTensorPtrOp>()) {
+        if (visitOperandMakeTPtr(makeTPtrOp, state, op.getLoc(), builder)
+                .failed()) {
+          newInitArgs.push_back(arg);
+          continue;
+        }
+        newInitArgs.push_back(mappedV);
+      } else if (auto addptrOp = mappedV.getDefiningOp<triton::AddPtrOp>()) {
+        assert(!addptrOp.getResult().getType().isa<RankedTensorType>());
+        if (visitOperandAddptr(addptrOp, state, op.getLoc(), builder)
+                .failed()) {
+          newInitArgs.push_back(arg);
+          continue;
+        }
+        newInitArgs.push_back(mappedV);
+      }
+    }
+    // Init arg is not pointer related or prior rewrite has failed. Pass as is
+    else {
+      newInitArgs.push_back(arg);
+      continue;
+    }
+    // Record the PtrSState for later processing
+    initArgIndexState.push_back(std::make_pair(i, state));
+  }
+
+  // For each of the PtrSState recorded in the last step, insert new instructions
+  // to describe offset and stride for each dimension and append them to init
+  // args
+  for (auto [i, state] : initArgIndexState) {
+    // For each dimension, if the corresponding offset and stride is an
+    // integer attribute, create a constant value and append them at the
+    // end of init arg list.
+    for (auto [j, s] : llvm::enumerate(state.offsets)) {
+      auto sIntAttr = getIntAttr(s);
+      if (sIntAttr) {
+        auto constOp = builder.create<arith::ConstantOp>(
+            op.getLoc(), builder.getIndexAttr(sIntAttr.value()));
+        newInitArgs.push_back(constOp.getResult());
+        state.offsets[j] = constOp.getResult();
+      } else {
+        newInitArgs.push_back(s.get<Value>());
+      }
+    }
+
+    for (auto [j, s] : llvm::enumerate(state.strides)) {
+      auto sIntAttr = getIntAttr(s);
+      if (sIntAttr) {
+        auto constOp = builder.create<arith::ConstantOp>(
+            op.getLoc(), builder.getIndexAttr(sIntAttr.value()));
+        newInitArgs.push_back(constOp.getResult());
+        state.strides[j] = constOp.getResult();
+      } else {
+        newInitArgs.push_back(s.get<Value>());
+      }
+    }
+
+    if (state.getRank() == 0) {
+      assert(state.scalar);
+      newInitArgs.push_back(state.scalar);
+    }
```

Comment:
this code is new, can we have an example snippet describing the case? i'm not quite following why we have to push the scalar to new init args

Reply:
I believe this is to account for scenarios where the init arg is a scalar pointer. Will confirm and add comment.